crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1601 @@
+
+        let memory_prefetch_func = get_memory_prefetch_func(verbose);
+
+        Self {
+            paths,
+            file_info,
+            hive_parts,
+            predicate,
+            options,
+            cloud_options,
+            file_options,
+
+            config: Config {
+                // Initialized later
+                num_pipelines: 0,
+                metadata_prefetch_size: 0,
+                metadata_decode_ahead_size: 0,
+                row_group_prefetch_size: 0,
+            },
+            verbose,
+            physical_predicate: None,
+            projected_arrow_fields: Arc::new([]),
+            memory_prefetch_func,
+
+            row_group_data_task_init_permit: None,
+            row_group_data_rx_handles: Arc::new(tokio::sync::RwLock::new(None)),
+            row_group_decoder: None,
+            is_finished: AtomicBool::new(false),
+        }
+    }
+}
+
+mod compute_node_impl {
+
+    use std::sync::Arc;
+
+    use polars_expr::prelude::phys_expr_to_io_expr;
+
+    use super::super::compute_node_prelude::*;
+    use super::{Config, ParquetSourceNode};
+    use crate::async_primitives::wait_group::WaitGroup;
+    use crate::morsel::SourceToken;
+    use crate::utils::semaphore_wrap::WithSemaphorePermit;
+
+    impl ComputeNode for ParquetSourceNode {
+        fn name(&self) -> &str {
+            "parquet_source"
+        }
+
+        fn initialize(&mut self, num_pipelines: usize) {
+            self.config = {
+                let metadata_prefetch_size = polars_core::config::get_file_prefetch_size();
+                // Limit metadata decode to the number of threads.
+                let metadata_decode_ahead_size =
+                    (metadata_prefetch_size / 2).min(1 + num_pipelines).max(1);
+                let row_group_prefetch_size = polars_core::config::get_rg_prefetch_size();
+
+                Config {
+                    num_pipelines,
+                    metadata_prefetch_size,
+                    metadata_decode_ahead_size,
+                    row_group_prefetch_size,
+                }
+            };
+
+            if self.verbose {
+                eprintln!("[ParquetSource]: {:?}", &self.config);
+            }
+
+            self.init_projected_arrow_fields();
+            self.physical_predicate = self.predicate.clone().map(phys_expr_to_io_expr);
+
+            let (row_group_data_rx, row_group_data_task_handle) =
+                self.init_row_group_data_fetcher();
+
+            self.row_group_data_rx_handles
+                .try_write()
+                .unwrap()
+                .replace((row_group_data_rx, row_group_data_task_handle));
+
+            let row_group_decoder = self.init_row_group_decoder();
+            self.row_group_decoder = Some(Arc::new(row_group_decoder));
+        }
+
+        fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) {
+            use std::sync::atomic::Ordering;
+
+            debug_assert!(recv.is_empty());
```

Comment:
Could you do a pass through the code and replace all occurrences of `debug_assert` with real `assert`s, if the assert in question is cheap? This is an example of something that costs almost nothing but could make future bug reproduction / troubleshooting a lot easier if it also triggers for end users.